### ISSUE TITLE
Random not implemented

### ISSUE THREAD

**Issue by roccosette** [keeweb/keeweb#876]
Hi,

When i try to make something (create file, open local file), i get error "Random not implemented" in the console.
I'm trying to search on search engine without results...

Thank's for reading

**Comment by antelle** [comment]
Hi,
Which browser is it?

**Comment by antelle** [comment]
Wow, how is this possible? Could you please check is WebCrypto available?
Please open your console and check the result of `typeof crypto.getRandomValues`, it should be `function`.

**Comment by roccosette** [comment]
On my web browser ?

**Comment by antelle** [comment]
Yes, sure. You get an error in your browser, right?

**Comment by roccosette** [comment]
typeof crypto.getRandomValues
"function"

keeweb.domain.com/:1 Uncaught KdbxError {name: "KdbxError", code: "NotImplemented", message: "Error NotImplemented: Random not implemented"}

**Comment by antelle** [comment]
I don't understand it... This error [is returned](https://github.com/keeweb/kdbxweb/blob/master/lib/crypto/crypto-engine.js#L155) if webcrypto is not available, something went terribly wrong.

**Comment by antelle** [comment]
Is there a stack track in this error? Usually you can expand it. Probably there's not much interesting inside, but just in case...

**Comment by roccosette** [comment]
I used :
git clone -b gh-pages https://github.com/keeweb/keeweb.git
and put in my apache .conf :
<VirtualHost *:80>
    ServerAdmin domain@gmail.com
    ServerName keeweb.domain.com
    DocumentRoot /var/www/keeweb/

    <Location "/webdav">
        DAV On
        Options Indexes
    </Location>

    ErrorLog ${APACHE_LOG_DIR}/keeweb.domain.com-error.log

    # Possible values include: debug, info, notice, warn, error, crit,
    # alert, emerg.
    LogLevel warn

    CustomLog ${APACHE_LOG_DIR}/keeweb.domain.com-access.log combined
    # Enable this for template generation
</VirtualHost>

I deleted htaccess but i already trying to put him...

**Comment by antelle** [comment]
Another question: does app.keeweb.info work in this browser?

**Comment by roccosette** [comment]
Yes it works

(index):1 Uncaught KdbxError {name: "KdbxError", code: "NotImplemented", message: "Error NotImplemented: Random not implemented"}code: "NotImplemented"message: "Error NotImplemented: Random not implemented"name: "KdbxError"__proto__: constructor: ƒ Error()stackTraceLimit: 10arguments: (...)caller: (...)captureStackTrace: ƒ captureStackTrace()length: 1name: "Error"prototype: {constructor: ƒ, name: "Error", message: "", toString: ƒ}__proto__: ƒ ()[[Scopes]]: Scopes[0]message: ""name: "Error"toString: ƒ toString()arguments: (...)caller: (...)length: 0name: "toString"__proto__: ƒ ()[[Scopes]]: Scopes[0]__proto__: Object
random @ (index):1
getBytes @ (index):1
get @ (index):1
openDb @ (index):1
inputKeydown @ (index):1
dispatch @ (index):1
q.handle @ (index):1

**Comment by antelle** [comment]
> Yes it works

You mean, app.keeweb.info works (sorry for asking multiple questions, I messed it up)?

**Comment by roccosette** [comment]
I can create new file on app.keeweb.info, i get this error on my server.
No problem, i prefer a correct comprehension for helping

**Comment by antelle** [comment]
Probably there's an encoding issue. 
Could you please check it? E.g.:
```
curl https://app.keeweb.info | md5
curl your-server | md5
```

**Comment by roccosette** [comment]
your-server = my keepweb website ?

**Comment by antelle** [comment]
Yep, where you host it. If it works, hashes will be the same.

**Comment by roccosette** [comment]
sorry...

curl http://keeweb.domain.com | md5sum  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1409k  100 1409k    0     0  72.8M      0 --:--:-- --:--:-- --:--:-- 76.4M
4185c0f1a508b2897f95ea7a82f24298  -
curl https://app.keeweb.info | md5sum  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1409k    0 1409k    0     0  1541k      0 --:--:-- --:--:-- --:--:-- 1541k
4185c0f1a508b2897f95ea7a82f24298  -

Same md5

**Comment by antelle** [comment]
I don't know then. You get the same app, in your server and app.keeweb.info. You do exactly the same (or, is there a difference?), and you get different results. That's something weird.

**Comment by roccosette** [comment]
You can see it on keeweb.roccosette.com, if this can help you.

**Comment by roccosette** [comment]
I just set the public domain, i check it for you see the domain

**Comment by antelle** [comment]
Http? :) It won't work. Now I see. 
So, if you host it on https, it will be fixed. It looks like, Chrome disabled WebCrypto on HTTP, I'll add a warning and the app will refuse to start form now on.

**Comment by antelle** [comment]
Changed the title

**Comment by roccosette** [comment]
Have you a doc for make ssl with apache ?

**Comment by antelle** [comment]
I don't know much about Apache, but there are plenty of docs around, you can probably easily google it.

**Comment by roccosette** [comment]
It's up

**Comment by antelle** [comment]
Sorry what do you mean? It doesn't work on https for you?

**Comment by Snapy** [comment]
@roccosette You should check [this] (https://certbot.eff.org/#debianstretch-apache).

**Comment by Snapy** [comment]
@roccosette You should check [this](https://certbot.eff.org/#debianstretch-apache).

**Comment by roccosette** [comment]
It's works with ssl :)

**Comment by antelle** [comment]
Cool! Let's keep it open to show a human-friendly message.

**Comment by neon-JS** [comment]
Will this work on http at one time? I'm currently hosting the webapp in my local network where there's no easy way to get https, I guess.

**Comment by antelle** [comment]
No. Furthermore, the following version will stop working if launched via HTTP. You can use self-signed certificates though.

**Comment by dploeger** [comment]
@antelle So this is actually fixed, right? The new version shows the appSecWarnBody1/2 and appSecWarn messages, which should be enough, I guess.

**Comment by antelle** [comment]
Yes, new version will show an error.
